### PR TITLE
GitHub Actions running PyTest

### PR DESCRIPTION
**Pull request by GenericP3rson**
Added GitHub actions which automatically runs the tests, allowing for quicker review workflow. As of now, skips the test requiring IBM credentials and skips work-in-progress test.

Additional minors changes to the density files to remove undefined variables and incorrect indentation.

### CODE REVIEW

File: .github/workflows/functional_tests.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  pull_request:
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.7", "3.8", "3.9", "3.10", "3.11"]
```
Comment (GenericP3rson) [inline .github/workflows/functional_tests.yaml:17]: Note: testing Python 3.7-3.11. According to our readme, we support 3.7-3.9. Python 3.7 end of life was June 27th, 2023. We can modify this as needed though; feel free to recommend more/less python versions!

**Comment by Hanrui-Wang** [comment]
seems that the checks for 3.7 to 3.11 all fail?

**Comment by GenericP3rson** [comment]
Looks like the tests are failing on https://github.com/mit-han-lab/torchquantum/blob/main/torchquantum/pulse/hardware/hardware.py because of the indentation issues (a bit odd that it didn't fail on that with that in my fork). I'll look into update the file!

**Comment by Hanrui-Wang** [comment]
> Contributor

i have updated it in the main & dev branch

**Comment by Hanrui-Wang** [comment]
seems to be some lint issues

**Comment by GenericP3rson** [comment]
Yeah, as of now, the lint just checks for if a variable is undefined or if there's a syntax issue. It looks like the hardware folder wasn't in my fork, so hopefully we can sort that out. 

It's also a bit odd that Python 3.11 is failing on the dependencies; it managed to pass in my fork, so I wonder if there's something else different that I missed: https://github.com/GenericP3rson/torchquantum/actions/runs/6059735082/job/16443152430

**Comment by GenericP3rson** [comment]
I'll try to rebase the dev branch on my fork and see if I can replicate the errors locally and fix them.

**Comment by GenericP3rson** [comment]
We're getting to the tests now. It looks like it can't find qiskit's CCZGate. However, locally, the tests are passing with:

```
qiskit==0.44.1
qiskit-aer==0.11.0
qiskit-ibmq-provider==0.19.2
qiskit-terra==0.25.1
```

**Comment by GenericP3rson** [comment]
The CCZ gate was released on 0.39.0 (as opposed to 0.38.0), so bumped qiskit and the tests are passing now!

File: .github/workflows/functional_tests.yaml (hunk 1 above)
Comment (Hanrui-Wang) [reply .github/workflows/functional_tests.yaml:17]: in that case we can modify the support to 3.8-3.11

**Comment by Hanrui-Wang** [comment]
wonderful. Thanks a lot Shreya!

**Pull request merged**